### PR TITLE
regexPrefix KlasaClientOption for Defaulted non-overidden regex prefix

### PR DESCRIPTION
**Pull request by MrJacz**
this was suggested by me in #94 mainly because prefix KlasaClientOption doesn't support Regexp's because of SG, This is handy as it cannot be overridden meaning you can have `prefix` in KlasaClientConfig but that can be changed this cannot be changed which comes in handy :), P.S Comments on using `msg.content.match()` and not `regexp.exec()`

### CODE REVIEW

File: tutorials/GettingStarted.md [hunk 1]
```diff
@@ -36,10 +36,10 @@ new Client({
 
 | Name                       | Default                   | Type               | Description                                                                         |
 | -------------------------- | ------------------------- | ------------------ | ----------------------------------------------------------------------------------- |
-| **clientOptions**          | `{}`                      | Object             | These are passed directly to the discord.js library. They are optional.¹            |
-| **prefix**                 | `undefined`               | string/regex/array | The default prefix(es) when the bot first boots up.²                                |
+| **clientOptions**          | `{}`                      | Object             | These are passed directly to the discord.js library. They are optional.Â¹            |
```
Comment (Pandraghon) [inline tutorials/GettingStarted.md]: What's that?

Comment (MrJacz) [reply tutorials/GettingStarted.md]: Its a `1`

Comment (Pandraghon) [reply tutorials/GettingStarted.md]: Why did you add `Â`?

Comment (MrJacz) [reply tutorials/GettingStarted.md]: I blame vs code for that :(((

File: tutorials/GettingStarted.md [hunk 2]
```diff
@@ -37,9 +37,9 @@ new Client({
 | Name                       | Default                   | Type               | Description                                                                         |
 | -------------------------- | ------------------------- | ------------------ | ----------------------------------------------------------------------------------- |
 | **clientOptions**          | `{}`                      | Object             | These are passed directly to the discord.js library. They are optional.¹            |
-| **prefix**                 | `undefined`               | string/regex/array | The default prefix(es) when the bot first boots up.²                                |
+| **prefix**                 | `undefined`               | string/regex/array | The default prefix(es) when the bot first boots up.Â²                                |
```
Comment (Pandraghon) [inline tutorials/GettingStarted.md]: Here

File: tutorials/GettingStarted.md [hunk 3]
```diff
@@ -37,9 +37,9 @@ new Client({
...
 | **permissionLevels**       | `defaultPermissionLevels` | PermissionLevels   | The permission levels to use with this bot                                          |
-| **clientBaseDir**          | see below³                | string             | The directory where all piece folders can be found                                  |
+| **clientBaseDir**          | see belowÂ³                | string             | The directory where all piece folders can be found                                  |
```
Comment (Pandraghon) [inline tutorials/GettingStarted.md]: Here

File: tutorials/GettingStarted.md [hunk 4]
```diff
@@ -50,10 +50,11 @@ new Client({
 | **cmdPrompt**              | `false`                   | boolean            | Whether the bot should prompt missing parameters                                    |
 | **cmdEditing**             | `false`                   | boolean            | Whether the bot should update responses if the command is edited                    |
 | **cmdLogging**             | `false`                   | boolean            | Whether the bot should log command usage                                            |
-| **quotedStringSupport**    | `false`                   | boolean            | Whether the bot should default to using quoted string support⁴                      |
+| **quotedStringSupport**    | `false`                   | boolean            | Whether the bot should default to using quoted string supportâ´                      |
```
Comment (Pandraghon) [inline tutorials/GettingStarted.md]: Here

File: tutorials/GettingStarted.md [hunk 5]
```diff
@@ -50,10 +50,11 @@ new Client({
...
 | **typing**                 | `false`                   | boolean            | Whether the bot should type while processing commands.                              |
-| **readyMessage**           | see below⁵                | string/function    | readyMessage to be passed through to Klasa's ready event.                           |
-| **ownerID**                | see below⁶                | string             | The discord id for the user the bot should respect as the owner                     |
+| **readyMessage**           | see belowâµ                | string/function    | readyMessage to be passed through to Klasa's ready event.                           |
```
Comment (Pandraghon) [inline tutorials/GettingStarted.md]: Here

File: tutorials/GettingStarted.md [hunk 6]
```diff
@@ -50,10 +50,11 @@ new Client({
...
+| **ownerID**                | see belowâ¶                | string             | The discord id for the user the bot should respect as the owner                     |
```
Comment (Pandraghon) [inline tutorials/GettingStarted.md]: Here

**Comment by MrJacz** [comment]
@Pandraghon Stop spamming GitHub Jks but I know just my internet is shitting itself and getting this :(
https://i.imgur.com/uWIodpF.png

File: src/monitors/commandHandler.js [hunk 7]
```diff
@@ -43,6 +44,7 @@ module.exports = class extends Monitor {
 
 	getPrefix(msg) {
 		if (this.prefixMention.test(msg.content)) return { length: this.nick.test(msg.content) ? this.prefixMentionLength + 1 : this.prefixMentionLength, regex: this.prefixMention };
+		if (this.client.config.regexPrefix) return { length: msg.content.match(this.regexPrefix)[0].length, regex: this.regexPrefix };
```
Comment (MrJacz) [inline src/monitors/commandHandler.js]: I'm not sure if i use msg.content.match() or RegExp.exec() here

**Comment by MrJacz** [comment]
@kyranet the `.gitattributes` file changes all files to LF in VS Code so no

**Comment by MrJacz** [comment]
image

File: src/monitors/commandHandler.js [hunk 8]
```diff
@@ -43,6 +44,7 @@ module.exports = class extends Monitor {
 
 	getPrefix(msg) {
 		if (this.prefixMention.test(msg.content)) return { length: this.nick.test(msg.content) ? this.prefixMentionLength + 1 : this.prefixMentionLength, regex: this.prefixMention };
+		if (this.regexPrefix.test(msg.content)) return { length: msg.content.match(this.regexPrefix) ? msg.content.match(this.regexPrefix)[0].length : null, regex: this.regexPrefix };
```
Comment (bdistin) [inline src/monitors/commandHandler.js]: null is never a valid value for length

Comment (bdistin) [reply src/monitors/commandHandler.js]: also, if regexPrefix is not defined, you will get error cannot read property test of null...

File: typings/index.d.ts [hunk 9]
```diff
@@ -1,1362 +1,1363 @@
...
-		public commandMessageSweep: number;
-		public ready: boolean;
-		public methods: {
-			Collection: typeof Collection;
-			Embed: typeof MessageEmbed;
-			MessageCollector: typeof MessageCollector;
-			Webhook: typeof WebhookClient;
-			CommandMessage: typeof CommandMessage;
-			util: Util;
+		import {
```
Comment (bdistin) [inline typings/index.d.ts]: this pr should not have such wide touching on this file.

File: src/monitors/commandHandler.js (hunk 8 above)
Comment (MrJacz) [reply src/monitors/commandHandler.js]: Latest commit should fix that

**Comment by MrJacz** [comment]
index.d.ts file is now fixed if you would like to take a look again :D

File: src/monitors/commandHandler.js [hunk 10]
```diff
@@ -119,6 +121,7 @@ module.exports = class extends Monitor {
 		this.ignoreOthers = !this.client.user.bot;
 		this.prefixMention = new RegExp(`^<@!?${this.client.user.id}>`);
 		this.prefixMentionLength = this.client.user.id.length + 3;
+		this.regexPrefix = this.client.config.regexPrefix;
```
Comment (bdistin) [inline src/monitors/commandHandler.js]: no need to init this, this.client.config.regexPrefix is available when this monitor is constructed.

File: src/monitors/commandHandler.js [hunk 11]
```diff
@@ -43,6 +44,7 @@ module.exports = class extends Monitor {
 
 	getPrefix(msg) {
 		if (this.prefixMention.test(msg.content)) return { length: this.nick.test(msg.content) ? this.prefixMentionLength + 1 : this.prefixMentionLength, regex: this.prefixMention };
+		if (this.regexPrefix && this.regexPrefix.test(msg.content)) return { length: msg.content.match(this.regexPrefix) ? msg.content.match(this.regexPrefix)[0].length : 0, regex: this.regexPrefix };
```
Comment (bdistin) [inline src/monitors/commandHandler.js]: split into multi lines, faster to exec once, than test + match + match
```js
if (this.regexPrefix) {
  const results = this.regexPrefix.exect(msg.content);
  if (results) return { length: results[0].length, regex: this.regexPrefix };
}
```

**Comment by MrJacz** [comment]
@bdistin Requested changes done

**Comment by bdistin** [comment]
closes #94

**Comment by MrJacz** [comment]
Merge conflicts fixed :)

**Pull request merged**